Serialize algorithm

moxcodes opened the pull request:
## Proposed changes

Adds the ability to serialize the `AlgorithmImpl`, which represents the charm++ 'chare' associated with all of our parallel components.
This also enables the ability to load-balance any of our existing executables.

There are a few parts of this serialization that I was entirely unsure how to test, in particular the serialization of the AlgorithmImpl, the ConstGlobalCache, and the CmiNodeLock, as they seem irreducibly associated with a true charm++ runtime, rather than the action-testing framework. I've added a test yaml that ensures load-balancing is performed successfully in the Burgers executable, but I'd welcome suggestions for good ways of unit testing these serializations more directly.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies

- [x] #2324
- [x] #2368
- [x] #2413
- [x] #2555
- [x] #2612
- [x] #2613
- [x] #2616

moxcodes committed:
Add serialization for Algorithm (components)

moxcodes commented:
Based on the results from my branch, I think CI will fail to compile on this due to memory pressure, so I imagine this will also need to wait until after the DataBox changes.

moxcodes commented:
Squashed in a fix that ensures that the load-balancing genuinely works on multiple nodes for our executables, so this branch can be useful for testing out load-balancing. CI will probably still fail though.

fmahebert commented on src/Parallel/ConstGlobalCache.hpp:
```diff
@@ -169,6 +173,14 @@ class ConstGlobalCache : public CBase_ConstGlobalCache<Metavariables> {
       const CkCallback& callback) noexcept;
 
+  /// Entry method to set the proxy to the const global cache itself (needed to
+  /// serialize cache pointers). This should be only called once.
```
Could you clarify "cache pointers" here? Do you mean pointers to the const global cache, or do you mean pointers contained as part of the const global cache data.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -407,8 +407,7 @@ AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>::
           db::wrap_tags_in<Tags::FromConstGlobalCache, all_cache_tags>>>(
-      static_cast<const Parallel::ConstGlobalCache<metavariables>*>(
-          const_global_cache_),
+      const_global_cache_,
```
To confirm my understanding: the reason you're getting rid of the `const` here (and elsewhere in this commit) is to enable the setting of the self-proxy?

moxcodes replied:
The const needs to be dropped so that the pack and unpack of the `DataBox` and `AlgorithmImpl` work appropriatey, because when the pup is called, the object is already created and needs to set the data memberrs, but if they're const, it isn't able to do that.

fmahebert commented on src/Parallel/ConstGlobalCache.hpp:
```diff
@@ -211,6 +225,21 @@ void ConstGlobalCache<Metavariables>::set_parallel_components(
+void ConstGlobalCache<Metavariables>::set_self_proxy(
+    const self_proxy_type& self_proxy) noexcept {
+  if (not self_proxy_has_been_set_) {
+    self_proxy_ = self_proxy;
```
Should this branch of the `if` set `self_proxy_has_been_set_ = true;` ?

Relatedly, keeping track of this bool is mildly annoying ... an alternative might be to have a pointer to the self-proxy, and check if the pointer is nullptr or not. This is obviously annoying in other ways. What do you think?

moxcodes replied:
I think the best fix here is to switch to a `boost::optional` -- I'll change that along with the other self-proxy fixes

fmahebert commented on src/Parallel/ConstGlobalCache.hpp:
```diff
@@ -211,6 +225,21 @@ void ConstGlobalCache<Metavariables>::set_parallel_components(
+
+template <typename Metavariables>
+auto ConstGlobalCache<Metavariables>::get_self_proxy() noexcept {
+  return self_proxy_;
```
What should the behavior here be if `self_proxy_has_been_set_ == false` ? As currently coded, I'm thinking the return value may be nonsense?

fmahebert commented on tests/Unit/Evolution/Systems/Cce/Test_ScriPlusInterpolationManager.cpp:
```diff
@@ -50,13 +50,16 @@ void test_interpolate_quadratic() {
         random_vector *
             (1.0 + linear_coefficient * (i * 0.1 + time_vector) +
              quadratic_coefficient * square(i * 0.1 + time_vector)));
+    if (test_serialization and i == data_points / 2) {
```
Can you add a short comment explaining why the `i == ...` condition? Is this a special case? Or are you just interested in testing the serialization once out of the N loops?

fmahebert commented on src/Parallel/NodeLock.hpp:
```diff
@@ -71,3 +71,31 @@ constexpr inline void unlock(
+
+/// \ingroup ParallelGroup
+/// Serialization of the node lock for Charm++ (just locks at origin and gets a
+/// new one at the destination)
```
Trying to improve clarity... I suggest the parenthetical -> "(locks at origin, and creates a new lock at the destination)" or similar.

fmahebert commented on src/Parallel/NodeLock.hpp:
```diff
@@ -71,3 +71,31 @@ constexpr inline void unlock(
+  if (p.isUnpacking()) {
+    t = Parallel::create_lock();
+  } else if (p.isPacking()) {
+    // just for safety, this ensures that nothing that wants to grab the lock
```
Trying to improve clarity by avoiding the semi-vague "this ensures"... I suggest phrasing this first sentence as "For safety, grab the lock so that nothing else can access the lock until the component has been unpacked at its destination" or similar.

fmahebert commented on src/Utilities/BoostHelpers.hpp:
```diff
@@ -77,6 +77,9 @@ void pup(er& p, boost::variant<Ts...>& var) noexcept {  // NOLINT
   p | send_index;
   (void)std::initializer_list<char>{
       ::BoostVariant_detail::pup_helper<Ts>(index, p, var, send_index)...};
+  if(var.which() != send_index) {
```
Apply clang-format

fmahebert commented on src/Utilities/BoostHelpers.hpp:
```diff
@@ -77,6 +77,9 @@ void pup(er& p, boost::variant<Ts...>& var) noexcept {  // NOLINT
   (void)std::initializer_list<char>{
       ::BoostVariant_detail::pup_helper<Ts>(index, p, var, send_index)...};
+  if(var.which() != send_index) {
+    ERROR("boost variant did not transmit successfully.");
```
- I suggest "boost variant" -> "boost::variant" for clarity
- Can you help me understand this check? When would this go wrong (as in, what does it mean for the variant not to transmit successfully)? Presumably this only ever triggers when unpacking? Perhaps putting this in an "if p.isUnpacking" and adding a comment might help.

moxcodes replied:
I think I'll actually just drop this check from the commit -- I added it because I thought this would fail in certain edge cases, but now I'm pretty sure as written this check always passes.

fmahebert commented on src/Parallel/ConstGlobalCache.hpp:
```diff
@@ -342,12 +344,14 @@ namespace PUP {
   } else {
-    t->get_self_proxy().pup(p);
+    local_const_global_cache_proxy = t->get_self_proxy();
+    p | local_const_global_cache_proxy;
```
[not a change request] Help: how are the two new lines different from the old line?

Should the changes to this file be moved into the earlier commit for const global cache?

moxcodes replied:
I think these currently don't actually make a difference, but I made this change after re-reading the charm++ pup documentation which notes that because of how the operator overloads vs member functions could be overloaded -- i.e. if we want to set the `self_proxy_type` based on template arguments to a type that doesn't have a pup member, we have to use the `p | local_const_global_cache_proxy;` syntax rather than `const_global_cache_proxy.pup(p);` syntax.

I agree, though, that this is in the wrong commit.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -174,6 +224,30 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>> {
+    if constexpr (Parallel::is_node_group_proxy<cproxy_type>::value) {
+      if (p.isUnpacking()) {
+        node_lock_ = create_lock();
+      }
```
Why not use `p | node_lock_` since you've made it serializable?

moxcodes commented:
fixups posted.
Cheers!

fmahebert commented:
Fixups LGTM. I think the conflicts are from the underlying PRs, so they are perhaps best resolved by waiting 😛 . I'll take another pass once everything is rebased and squashed.

moxcodes commented:
RIP I forgot to deal with un-registering and re-registering with observers.

moxcodes commented:
Okay the observer thing is a large enough change to need to be in follow-up PRs, so I'm removing the 'in-progress' tag

moxcodes commented:
Thanks for the review @fmahebert ! I've posted some fixups -- the changes to `Algorithm.hpp` were sufficiently large that I just squashed them in to the `Add global sync phase control to algorithm`. I also added a documentation commit.

nilsdeppe commented on src/Parallel/NodeLock.hpp:
```diff
@@ -71,3 +71,31 @@ constexpr inline void unlock(
     const gsl::not_null<NoSuchType*> /*unused*/) noexcept {}
 /// \endcond
 }  // namespace Parallel
+
```
Please drop the serialization of the NodeLock. There's a good reason Charm++ doesn't offer it:
```cpp
src/conv-core/converse.h:typedef int CmiNodeLock;
```
it's just an `int` or some other weird type depending on various `ifdef`s. Furthermore, it makes absolutely no sense to send a lock since locks are tied to hardware. The weird code inside the `pup` function also makes it clear that the lock should never be serialized. Additionally, the lock is never destroyed/freed, which makes this code even more difficult to use.

moxcodes replied:
Right, so I agree the serialization of the node lock is weird, and my attempt at an implementation can probably be improved, but hear me out -- we _do_ need to serialize objects that contain the NodeLock, because even the groups and node groups get serialized when checkpointing. And when those objects get unpacked, we need to set up the node lock to a usable state.

So, an alternative that would be reasonable (and I started trying before I did this) is to just move any of the node-lock manipulation into the serialization for each separate structure that contains them. It's probably not many, just the `AlgorithmImpl` and the `DataBox`. But, I'm not sure how to approach the serialization there any differently -- it seems as though we should definitely create a new lock when unpacking, and we should probably not free it when packing because (if I understand correctly) it is incorrect to destroy packed objects, instead letting any free occur on destruction.

So, my reasoning was that if we have the same strategy for all cases where we pup something that has the node lock then it may as well be in a centralized location for pupping the node lock, but if it's a preferable design to move it to the objects that contain node locks, I'm fine with that.
Is this misguided, what do you think?

nilsdeppe replied:
I had forgotten we store locks in the DataBox, which complicates things. My concern is actually primarily because of the way Charm++ implements the locks using type aliases. The reason being is that it'll be nearly impossible to avoid adding weird PUP overloads that could cause compiler errors on some architectures. Since we already wrap the locking, unlocking, creation, and destruction I think the best way forward would be to do https://github.com/sxs-collaboration/spectre/issues/998 and write a `Parallel::NodeLock` class that wraps the Charm++ node lock and thus is strongly typed eliminating my concerns. While serializing it is still weird, it then starts to become more reasonable. We can also make it so that the default constructor acquires the lock and the destructor frees the lock. E.g.
```cpp
namespace Parallel {
class NodeLock {
  NodeLock() noexcept {
    lock_ = CmiCreateLock();
  }

  NodeLock(const NodeLock&) = delete;
  NodeLock& operator=(const NodeLock&) = delete;
  NodeLock(NodeLock&&) = default;
  NodeLock& operator=(NodeLock&&) = default;

  ~NodeLock() noexcept {
    CmiDestroyLock(lock_);
  }

  void lock() noexcept {
    CmiLock(lock_);
  }

  bool try_lock() noexcept {
    return CmiTryLock(lock) == 0;
  }

  void unlock() noexcept {
    CmiUnlock(lock_);
  }

  void pup(PUP::er& p) noexcept {}

 private:
  CmiNodeLock lock_{};  // NOLINT
};
}  // namespace Parallel
```

Do you think that would solve the problem?

moxcodes replied:
Ah, yeah gotcha -- I agree, then, that the best thing to do is to just add the wrapper class. I especially like that this will do the 'right thing' of always destroying the lock on class destruction, analogous to good smart pointer practice.
I'll add the class as you suggest and replace existing node lock usage with the wrapper in a separate PR (on which this one will then depend) ~and switch the pup function here to act on the wrapper~ (edit: the pup function should probably just also be in that other PR).
Thanks, @nilsdeppe !

nilsdeppe replied:
That sounds great. Thank you!

I think you only need the member `pup` function in the `NodeLock` class, since that's what we do with all our other classes. There are templates that act on classes with `pup` member functions. Or maybe I'm misunderstanding what you mean by switching the pup function to act on the wrapper.

nilsdeppe replied:
Oh, I think the `move` semantics should do a `CmiCreateLock` and `CmiDestroyLock` since it's not clear that the underlying locks can be moved in any reasonable way, but I think moving a lock in general is well-defined

moxcodes commented:
squashed in the fixups, now dependent on the nodelock wrapper PR #2413 , and I dropped the original commits from this PR that dealt with the nodelock. I also dropped the commit that added a load-balancing test yaml because with the observers not being handled correctly in this PR, having the test that makes periodic LB strategies _seem_ to work is misleading at best.

moxcodes commented:
rebased on develop, no longer dependent.
Cheers!

moxcodes commented:
... except now the build is failing in a way I cannot reproduce locally. :/

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -21,6 +21,8 @@
 #include "DataStructures/DataBox/TagTraits.hpp"
 #include "ErrorHandling/Error.hpp"
 #include "ErrorHandling/StaticAssert.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
```
why are these being added here?  This is causing the gcc-9, Debug, OFF failure...

moxcodes replied:
If I recall correctly, I needed to add these so that the databox serialization found the new pup functions for the const global cache pointer and std::atomic. Could you help me understand what's going wrong with the dependencies? Maybe there's an easy workaround, like separating out the const global cache pointer pup to a file other than `ConstGlobalCache.hpp`.

kidder commented:
any  target that depends upon an include of  `Parallel/ConstGlobalCache.hpp` must have a dependency on `module_ConstGlobalCache` to make sure the charm generated includes are built first. The ConstGlobalCache is not mentioned in `DataBox.hpp` so we shouldn't have it included.

moxcodes commented:
ah, understood, thanks @kidder ! I've removed the includes, hopefully I was just mistaken about them being necessary for something in the dependency tree.

kidder commented on src/Domain/Creators/TimeDependence/RegisterDerivedWithCharm.cpp:
```diff
@@ -26,9 +25,12 @@ struct get_maps {
+      tmpl::remove_duplicates<tmpl::flatten<tmpl::push_back<
+          tmpl::transform<typename TimeDependence<Dim>::creatable_classes,
+                          get_maps<tmpl::_1>>,
+          domain::CoordinateMap<Frame::Grid, Frame::Inertial,
```
I think this addition is causing the clang-9, Debug, ON build failure, (which is built with shared libs) because it requires this file to need an instantiation of `domain::CoordinateMap<...>` which requires the include of `CoordinateMap.tpp`

moxcodes committed:
Add missing includes

- Test_LeviCevitaIterator.cpp
- ArrayIndex.hpp

moxcodes commented:
Now dependent on #2555, but the good news is that the optimizations in that PR are enough to get CI to run to completion on this one.

moxcodes commented:
Rebased on develop, no longer dependent.
Cheers!

fmahebert commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -306,6 +319,38 @@ struct FromGlobalCache : CacheTag, db::ComputeTag {
+namespace PUP {
+
+/// \ingroup ParallelGroup
+/// Serialization of a pointer to the const global cache for Charm++
```
Remove "const"

fmahebert commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -306,6 +319,38 @@ struct FromGlobalCache : CacheTag, db::ComputeTag {
+                Parallel::GlobalCache<Metavariables>*& t) noexcept {
+  boost::optional<
+      typename Parallel::GlobalCache<Metavariables>::self_proxy_type>
+      local_const_global_cache_proxy;
```
Rename `local_const_global_cache_proxy` -> `local_global_cache_proxy` ?

fmahebert commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -306,6 +319,38 @@ struct FromGlobalCache : CacheTag, db::ComputeTag {
+}
+
+/// \ingroup ParallelGroup
+/// Serialization of a pointer to the const global cache for Charm++
```
Remove "const"

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -46,6 +56,8 @@ namespace Parallel {
 struct Nodegroup;
 struct Singleton;
+struct Group;
+struct Array;
```
[nitpick] alphabetize these 4 lines

fmahebert commented on cmake/SetupCharm.cmake:
```diff
@@ -28,7 +28,7 @@ endif(NOT TARGET Charmxx)
 string(
     REGEX REPLACE "<CMAKE_CXX_COMPILER>"
-    "${CHARM_COMPILER} -pthread -module DistributedLB -no-charmrun"
+    "${CHARM_COMPILER} -pthread -module CommonLBs -no-charmrun"
```
Unclear to me if we want to merge this commit now as part of this PR, or later after the registration PR when more pieces are in. Thoughts?

nilsdeppe replied:
I think this was mainly a stub/placeholder so I have no objection to merging now

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -400,15 +400,14 @@ AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>::
-          tmpl::list<Tags::GlobalCacheImpl<metavariables>,
+  box_ =
+      db::create<db::AddSimpleTags<tmpl::flatten<tmpl::list<
+                     Tags::GlobalCacheImpl<metavariables>,
```
Can I request that you split the first commit "Add serialization for GlobalCache" into two commits, one for the whitespace changes and one for the "real" changes? There's a bit of noise in the diffs right now. I think the other commits are fine :)

fmahebert reviewed (commented):
Here are a few minor comments after looking through this again. Please squash directly. What's your strategy for the mutable cache that just got merged... expand this PR or do it in a future PR?

moxcodes commented:
Ah, yeah, I'll add a commit to this PR for serializing the mutable global cache as well.

nilsdeppe commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -37,8 +42,9 @@ struct type_for_get_helper<std::unique_ptr<T, D>> {
-  using type = tmpl::filter<ListOfPossibleTags,
-               std::is_base_of<tmpl::pin<GlobalCacheTag>, tmpl::_1>>;
+  using type =
+      tmpl::filter<ListOfPossibleTags,
```
I'm not sure if you run clang format on the entire file or why there are so many white space changes. `git clang-format -f` works nicely and only formats lines you actually touch. This makes the review a lot easier so I don't spend time trying to figure out what changed when the answer is "nothing"

kidder commented on src/Parallel/Algorithm.hpp:
```diff
@@ -174,6 +225,28 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>> {
+    p | terminate_;
+    p | box_;
+    p | inboxes_;
+    if constexpr (not std::is_same_v<typename ParallelComponent::chare_type,
```
there is no reason not to serialize array_index_ for a singleton (it's just a byte)

kidder commented:
Please break this up into several PRs with 
- clearer commit messages
- no reformatting of lines you didn't change within commits that change code.

moxcodes commented:
Thanks @kidder -- I agree, this PR was previously a bit of a jumble.
I've stripped out two of the serialization additions to PRs #2612 and #2613, and moved some of the rest of the alterations that were necessary to compile with load-balancing to a branch for a future pull request.
Cheers!

nilsdeppe commented:
I guess this just needs a rebase mainly?

nilsdeppe commented on src/Parallel/Algorithm.hpp:
```diff
@@ -182,6 +175,28 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
   /// Charm++ migration constructor, used after a chare is migrated
   constexpr explicit AlgorithmImpl(CkMigrateMessage* /*msg*/) noexcept;
 
+  void pup(PUP::er& p) noexcept override {  // NOLINT
```
Is there some way we can test this? For example in the `Test_Algorithm*` executables?

moxcodes replied:
So, it's a little tricky to test the serialization manually in those tests because the algorithm tests are actually run in the charm runtime.
What we could do easily is enable the CommonLBs compiler flag (either for those algorithm tests, or just generally), and have an algorithm test run with e.g. `RandCentLB` that will always migrate, and confirm that the output has evidence of migration from LBDebug.

fmahebert replied:
A checkpoint-restart based case might be nicer, in that it could test the different kinds of chares? Not helpful now, but could be added in a few PRs after this one.

nilsdeppe replied:
Would it not be possible to grab the `local()` in an action and then serialize/deserialize and test everything in the deserialized is correct or would this confuse Charm++? Or would having a test where you construct just an AlgorithmImpl and serialize/deserialize work and be easier? Generally when we've had bugs they've been subtly hidden in `pup` functions and this seems like one of the more important pup functions :P

moxcodes replied:
Right, so I think we could do that (I don't know if we've ever manually serialized an object that inherits from `CBase_blahblah` in a test, but I don't see anything going immediately wrong) -- the main challenge is then how we examine all of the things that we claim to be `pup`ing.
I agree this needs to be tested well because bugs in this will be subtle and affect effectively every executable. 

We could create an equivalence operator that verifies each of the member variables are the same before and after pupping. This is not terribly convincing, because all it really tests is that we forgot the same member variables in both the pup and the operator==...

Ideally, we'd have a test where the packed and unpacked algorithm is actually used to run the rest of the test, but I'm not sure how to do that without going through the charm runtime. Eventually, we'll need to have the ability to dynamically add and remove elements, but our phase structure currently can't do that I think.

We could add utilities to expose manual migration to the algorithm if that's more trustworthy than a balancer, then confirm that the element has actually been moved to a new processor. The one part of that version or the balancer version that's a bit unclear to me is whether it is guaranteed that charm will actually universally call the pup function when moving between cores on the same node. We could verify that last point by adding an extra debugging member function that returns the current memory address of the databox or something to verify that data has actually been moved.

What do you think?

moxcodes committed:
Add base class construction with CkMigrateMessage

- to GlobalCache, MutableGlobalCache, and AlgorithmImpl

moxcodes commented:
I've pushed a slightly creative version of a test that is a bit inelegant, but does at least verify that the pupping is happening and that the algorithm can continue running afterwards.
Let me know how you feel about it, and whether you'd prefer an expanded/altered test.
Cheers!

nilsdeppe commented on tests/Unit/Parallel/CMakeLists.txt:
```diff
@@ -110,6 +110,37 @@ function(add_algorithm_test_with_input_file TEST_NAME INPUT_FILE_DIR)
+    COMMAND
+    ${SHELL_EXECUTABLE}
+    -c "${CMAKE_BINARY_DIR}/bin/Test_${EXECUTABLE_NAME} +p2 +balancer \
+     RandCentLB +LBPeriod 5.0e-5 +LBDebug 3 2>&1"
```
I have a couple concerns with this test.
- This won't work with C++ modules (whenever we get to switch to them) because modules requires that you don't switch definitions locally based on macros. We can do this as a temporary hold over until more infrastructure gets in, but basically this can't really stay in the code very long otherwise it'll get forgotten and make somebody's life really terrible when they try to switch the code to modules.
- Why not use the `RotateLB`? The description for that is: `This strategy moves objects to the next available PE every time it is called. It is useful for debugging PUP routines and other migration related bugs.` and that's exactly what we want to do here: debug/test PUP routines. We could stick something in the DataBox and Inbox, then make sure things get migrated correctly. I think we can test the phase, databox, algorithm step, and inboxes (maybe the array_index too?) via some action calls. I'm not sure how to test the global cache serialization without going to multinode, but we can at least verify that we can retrieve a correct value out of the cache both explicitly and through the DataBox. 
I don't know how much of the suggested test can be done with what's already in the code, so this is mainly a discussion point to see what we can do now and what we ultimately I think should do for testing this.

moxcodes replied:
Gotcha, that's a good point regarding modules. The main problem I'm trying to solve with that trick is 'how do we trust that charm has really called the pup function'. I think the only way we can truly verify that is to extract some debugging information from the pup function or from extra debug functions in the AlgorithmImpl, or by manually running the serialization ourselves.

Regarding `RotateLB` - I tried that first, but I couldn't manage to get it to consistently migrate on two cores. I can't tell if that's a charm problem or a failure in my understanding.

Here's an alternative idea that shouldn't be too painful to code up -- in the `Test_AlgorithmParallel.cpp` test, add an action for the array that when the number of processors is exactly one (so we can be sure that other elements on the same core won't be simultaneously running), manually serializes/deserializes every other element of the array in circular permutation, then continues with the rest of the algorithm test. Does that sound robust enough to not need to be rewritten in the near future?

The only other note there is that it requires that at least one case of `Test_AlgorithmParallel.cpp` to be run on a single core, though I think that's how it works in the current cmake now anyways?

nilsdeppe replied:
Uh oh, this is spooky... Does `RotateLB` work on more than two cores? I think that actually needs to be understood because if we're seeing "weird load-balancing behavior" we should at least be able to do a rotate of the elements across cores correctly independent of the number of cores. If there is an issue on a specific number of cores we should be sure it's not how we use LBs because if we encounter other issues than we can't be sure why.

I think doing the explicit serialize/deserialize in `Test_AlgorithmParallel.cpp` should work. You're saying get the local object, serialize/deserialize and then do what with it? This seems like it could work, but I think we should run it through the sanitizers to make sure Charm doesn't make some assumption that we violate

moxcodes committed:
Make DataBox type error more verbose

moxcodes committed:
Fix bugs in Test_AlgorithmParallel

- previously the Array algorithm wasn't actually run
  (because the `perform_algorithm` should have been `start_phase`)
- previously the terminate was called with the DataBox in a
  state that additional phases could not be added after the
  PerformAlgorithmArray phase.

moxcodes commented:
Okay, I've pushed another try at the pup diagnostic that I think is a bit better (at least it doesn't use macros...).

To answer @nilsdeppe 's question about `RotateLB` - the correct answer is that it does actually migrate, but reports doing so differently in the LBDebug than the other LB's so I didn't recognize that it was doing anything on my first pass.

I tried the explicit serialize/deserialize, and the pup worked, but (slightly hilariously) the components then stored internally a different array_index than the charm runtime thought they had, so the components ended up very confused.

The current test verifies that the serialize/deserialize occurs several times, and that the test proceeds unimpeded, which I think is a reasonable verification that the information is transmitted successfully.

I've added commits:
```
Make DataBox type error more verbose
Fix bugs in Test_AlgorithmParallel
```
as well as changed the `fixup`s. The `DataBox type error` commit isn't strictly needed for this PR, but I've been frustrated with that error message a number of times, and when debugging this test I finally decided to actually improve matters.

moxcodes commented:
Sorry, there were a few things that broke in surprising ways on the last attempt -- now the fixups should be ready for another look.
I had to add a default constructor and use it when we construct the caches manually, because constructing with the CkMigrateMessage manually is no good (presumably it needs to be something specific from the charm runtime to work properly when it correctly passes to the `CBase...` classes).

kidder reviewed (commented):
I am fine with the fix-ups.  @nilsdeppe please look at

nilsdeppe commented on tests/Unit/Parallel/CMakeLists.txt:
```diff
@@ -110,6 +110,37 @@ function(add_algorithm_test_with_input_file TEST_NAME INPUT_FILE_DIR)
       ENVIRONMENT "ASAN_OPTIONS=detect_leaks=0")
 endfunction()
 
+# Note that we have a 'balance' here that happens 2.5e4 of times per second just
```
so I'm clear and that it's clearly documented: if the test were to run less than 4e-5 seconds then no balancing would occur? It might be good to note that explicitly in case we get failures, including as a C++ comment by the check. Just so that if some day we have sufficiently fast CPUs (very unlikely) we won't be totally baffled

moxcodes replied:
It depends on the run, there's a lot of noise in how many balances occur -- if you balance much less than this, I found that I would not consistently get the 5 balances I demand in the test. On average, this gives several dozen balances on my machine, it's just the tail of the distribution can get low sometimes. Too much more often balancing than this and the test takes super long because it's balancing period is about as long as the time it takes to balance, so it doesn't get much time to take computational action :P.
I'll add a bit more explanation to the cmake comment to that effect.

nilsdeppe commented on tests/Unit/Time/StepChoosers/Test_Cfl.cpp:
```diff
@@ -55,7 +55,7 @@ struct Metavariables {
                       const double characteristic_speed,
                       const DataVector& coordinates) noexcept {
-  const Parallel::GlobalCache<Metavariables> cache{{}};
+  const Parallel::GlobalCache<Metavariables> cache{};
```
Just for my understanding (and then maybe a suggestion): this calls a contructor with a `nullptr` secretly? And then what sort of an error does one get? If it's not clear could be add an 'ERROR' in a constructor to make things easier to debug? I don't know what the current error was, so I'm just checking that it is reasonably easy to debug :)

moxcodes replied:
Right, so my understanding from the error was that it was calling the `CkMigrateMessage*` constructor with a `nullptr`, and doing so was fine when we were forgetting to pass to `CBase...`, but the `CBase...` constructors error with some kind of terrible segfault when you hand them a malformed migrate message, so ... not super easy to debug.
I'll add an `ERROR` for the `nullptr` case and a documentation note. It doesn't actually catch all the problems, because even a valid pointer to an incorrectly constructed message object also led me to a segfault :(.

nilsdeppe reviewed (commented):
LGTM! I like this test :) One of the only places where I think we actually need to run a test on more than one core :) 

I have a couple small suggestions you can make (if they are reasonable) while squashing, no need for another fixup

moxcodes commented:
fixups squashed in, along with the additional `ERROR`s and code comments in response to @nilsdeppe 's last comments.
Cheers!

kidder commented:
@nilsdeppe please look at

nilsdeppe commented on src/Evolution/Systems/Cce/CMakeLists.txt:
```diff
@@ -54,6 +54,7 @@ spectre_target_headers(
 target_link_libraries(
   ${LIBRARY}
   PRIVATE
+  CceAnalyticSolutions
```
This is also in public a few lines below now (I had to add it when to get shared libs working in CI). I don't remember how carefully I check whether or not it needed to be public or private.

moxcodes commented:
dropped the linker fix that has apparently already been done on `develop`
Cheers!
